qiskit/algorithms/minimum_eigensolvers/vqe.py

```diff
@@ -0,0 +1,379 @@
+
+    The optimizer can either be one of Qiskit's optimizers, such as
+    :class:`~qiskit.algorithms.optimizers.SPSA` or a callable with the following signature:
+
+    .. code-block:: python
+
+        from qiskit.algorithms.optimizers import OptimizerResult
+
+        def my_minimizer(fun, x0, jac=None, bounds=None) -> OptimizerResult:
+            # Note that the callable *must* have these argument names!
+            # Args:
+            #     fun (callable): the function to minimize
+            #     x0 (np.ndarray): the initial point for the optimization
+            #     jac (callable, optional): the gradient of the objective function
+            #     bounds (list, optional): a list of tuples specifying the parameter bounds
+
+            result = OptimizerResult()
+            result.x = # optimal parameters
+            result.fun = # optimal function value
+            return result
+
+    The above signature also allows one to directly pass any SciPy minimizer, for instance as
+
+    .. code-block:: python
+
+        from functools import partial
+        from scipy.optimize import minimize
+
+        optimizer = partial(minimize, method="L-BFGS-B")
+
+    Attributes:
+        estimator (BaseEstimator): The estimator primitive to compute the expectation value of the
+            circuits.
+        ansatz (Quantum Circuit | None): A parameterized circuit, preparing the ansatz for the wave
+            function. If provided with ``None``, this defaults to a :class:`.RealAmplitudes`
+            circuit.
+        optimizer (Optimizer | Minimizer): A classical optimizer to find the minimum energy. This
+            can either be a Qiskit :class:`.Optimizer` or a callable implementing the
+            :class:`.Minimizer` protocol.
+        gradient (BaseEstimatorGradient | None): An optional estimator gradient to be used with the
+            optimizer.
+        initial_point (Sequence[float] | None): An optional initial point (i.e. initial parameter
+            values) for the optimizer. If ``None`` then VQE will look to the ansatz for a preferred
+            point and if not will simply compute a random one.
+        callback (Callable[[int, np.ndarray, float, float], None] | None): A callback that can
+            access the intermediate data during the optimization. Four parameter values are passed
+            to the callback as follows during each evaluation by the optimizer for its current set
+            of parameters as it works towards the minimum. These are: the evaluation count, the
+            optimizer parameters for the ansatz, the evaluated mean and the evaluated standard
+            deviation.
+
+    References:
+        [1] Peruzzo et al, "A variational eigenvalue solver on a quantum processor"
+            `arXiv:1304.3061 https://arxiv.org/abs/1304.3061>`_
+    """
+
+    def __init__(
+        self,
+        estimator: BaseEstimator,
+        ansatz: QuantumCircuit | None,
```

Comment:
Hi Declan! We had another discussion about defaults yesterday, and @mrossinek proposed dealing with the ansatz issue internally in UCCSD (etc) factories. This way we would not need to support type `None` as a type. I think this is a good idea, as this is not the way the algorithm is supposed to be used by "end-users" anyway.

Reply:
Ok, great. That's certainly a cleaner interface :).